### PR TITLE
Added skipDestructiveCodeActions argument to organize imports server command

### LINKED ISSUE #43051: source.organizeImports should not remove imports if file contains errors.
# Bug Report

When a user has `source.organizeImports` enabled on save/blur, the imports are mistakenly removed if the file contains errors. It also happens file contains an unclosed block comment `/*` (for example, when the file got saved in the middle of writing a comment).

### 🔎 Search Terms

- Remove unused imports
- source.organizeImports

### 🕗 Version & Regression Information

`typescript@4.1.2`

VSCode:
```
Version: 1.53.2
Commit: 622cb03f7e070a9670c94bae1a45d78d7181fbd4
Date: 2021-02-11T11:45:54.515Z (2 wks ago)
Electron: 11.2.1
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Darwin x64 19.6.0
```

- This is a crash: No
- This changed between versions ______ and _______: I do not recall this ever working
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because do not have an environment setup.

### ⏯ Playground Link
- I do not know how to trigger `source.organizeImports` in a playground.

### 💻 Code

An example of a broken file. This example contains JSX, hover, the point here if the file IS broken, then we should not remove imports.

File: `test.ts` (NOT `test.tsx`)
```text/typescript
import { uniq } from "lodash";

export default function useAssetTypeFilterOptions() {
  return (
    <input
      name="aaaaaaaaaaaaaaaaaaaaaaaaaaaa"
      value={uniq(["a", "b", "c"]).join(" + ")}
    />
  );
}
```

### 🙁 Actual behavior

- The imports are erroneously removed

### 🙂 Expected behavior

- The imports should not be removed if the source file contains errors.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Fixes #43051

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
I think this is *probably* okay, but part of me feels like there's a difference between a user-specified import cleanup and one that was triggered incidentally by organize-imports-on-save.

**Comment by JoshuaKGoldberg** [comment]
Err, does such a preference exist in the services API? I don't see anything in `src/compiler/types.ts` but could add it that consideration if available.

**Comment by andrewbranch** [comment]
The combination of codefixes on save and save on blur is so horrifying to me; to pick up on Daniel’s point, I think it merits a broader discussion about what kinds of actions should be allowed through that sequence. My feeling is we should never run anything that deletes code without an explicit user action, and blurring your editor tab doesn’t count as an explicit user action.

**Comment by jessetrinity** [comment]
Doesn't "do X on save" require that the use take explicit action in the first place (add the configuration to vscode settings.json)?

**Comment by andrewbranch** [comment]
Yeah, but I’m talking about direct synchronous action. I’m not saying “this is dangerous so it needs to be opt-in”; I’m saying “we should not have a feature that deletes code when users aren’t looking, no matter how badly they think they want it.”

**Comment by andrewbranch** [comment]
And to clarify, I’m not talking about codefixes-on-save alone. That’s ok because the direct causal action is Ctrl+S. If it messes up, you’re actively observing the mistake and you can undo it. The problem is save-on-blur _triggering_ codefixes-on-save. Because in that case, the you’re explicitly moving your attention _away_ from the place where we’re performing the edits. Anything done via that chain reaction needs to be totally non-destructive (e.g., formatting).

**Comment by jessetrinity** [comment]
Thanks for the clarification. Does this make sense as a user preference, something like "remove unused imports when organizing imports"?

**Comment by jessetrinity** [comment]
@JoshuaKGoldberg we discussed this a bit with VSCode and we think it makes sense to add a user preference to control if code is deleted on save.

I propose `allowDestructiveCodeActions?: boolean`. Default would be `true` to keep the current behavior. 'false' would short circuit in the same way as your syntax fix. VSCode could decide whether to pass `true` or `false` based on how the code action is triggered.

The syntax error would then be checked after this option.

**Comment by JoshuaKGoldberg** [comment]
Super. Is that something I'd be well positioned to fix? I'm wary of taking on big projects that are still in early stages cross-team as an external contributor, but if you have a solid vision I could contribute to that'd make me very happy.

**Comment by jessetrinity** [comment]
The work on our (TypeScript) side should be pretty simple. You would just need to add the option to the existing `UserPreferences` in `src\compiler\types.ts` and `src\server\protocol.ts`. `organizeImports` already takes a `UserPreferences` argument, so you would just need to then check the new option in the code removal logic. I think you would also need to change the [fourslash method]( https://github.com/microsoft/TypeScript/blob/ffe925ddd5ed8174ed6634fb16368fe2e4a2e7e3/src/harness/fourslashImpl.ts#L530) to test the new behavior.

You don't necessarily need to make changes on the vscode side to get this change in but you're welcome to try. If you're not comfortable doing that we can just let them know about the new option.

**Comment by andrewbranch** [comment]
Note that `UserPreferences` are changed in a separate `Configure` request, so I’m not sure that (by itself) actually plays well with the idea of letting VS Code change it on the fly depending on how a command is triggered. We might need to add it as an argument to other commands.

**Comment by jessetrinity** [comment]
Oh right. So for this particular issue, `OrganizeImportsRequestArgs`.

**Comment by JoshuaKGoldberg** [comment]
Thanks!  3a61904 adds the option here. I've dabbled in VS Code before; it'd be fun to try to add it there too. 😄

File: src/services/organizeImports.ts
Comment (andrewbranch) [inline src/services/organizeImports.ts:89]: I think the consensus from the meeting we had was that we should hold off on considering syntax errors until we’ve  seen that other mitigations aren’t sufficient. But @jessetrinity can correct me if I’m wrong.

Comment (JoshuaKGoldberg) [reply src/services/organizeImports.ts:89]: That makes it easier for me 🤠

**Comment by JoshuaKGoldberg** [comment]
Btw, I realized that 3a61904 (which added the `allowDestructiveCodeActions` argument) is a breaking change by nesting the `scope` part of the args. So in ~7fd22f8~ 1637e1a I switched the change to be an extension, not a nesting.

**Review by andrewbranch — APPROVED** [review]

**Comment by andrewbranch** [comment]
Thanks @JoshuaKGoldberg! Let’s get a review from @jessetrinity too.

**Review by jessetrinity — COMMENTED** [review]
nit: a word

**Review by jessetrinity — APPROVED** [review]
Looks good. We should make a separate issue for adding a `UserPreference` for cases in which the command is explicitly requested.

**Comment by andrewbranch** [comment]
Why didn’t @typescript-bot issue the pinging-everyone-because-you-changed-protocol.ts comment?

File: src/server/protocol.ts
Comment (amcasey) [inline src/server/protocol.ts:682]: I'm confused by this change.  What happened to the `scope` property?

File: src/server/protocol.ts
Comment (amcasey) [inline src/server/protocol.ts:683]: To confirm, we expect the editor to want to make this decision per-request and that's why it's not a UserPreference?

**Review by amcasey — CHANGES_REQUESTED** [review]
Seems like a breaking change to the protocol?  Am I missing something?

**Comment by amcasey** [comment]
> Btw, I realized that [3a61904](https://github.com/microsoft/TypeScript/commit/3a619048598fd446dcc789c082ffc5d5b0791b2b) (which added the `allowDestructiveCodeActions` argument) is a breaking change by nesting the `scope` part of the args. So in ~[7fd22f8](https://github.com/microsoft/TypeScript/commit/7fd22f8f737b58f5a3896e86d4301031f9203b2b)~ [1637e1a](https://github.com/microsoft/TypeScript/commit/1637e1ad4277b775fe30c42b230c35b1b9cb81f3) I switched the change to be an extension, not a nesting.

I'm definitely missing something in this comment.

File: src/server/protocol.ts
Comment (sheetalkamat) [reply src/server/protocol.ts:683]: Also i would name this flag as `skipDestructiveCodeActions` so that you would need to check only for truthiness instead of defaults when not set..

Comment (jessetrinity) [reply src/server/protocol.ts:683]: >To confirm, we expect the editor to want to make this decision per-request and that's why it's not a UserPreference?
Yes, the editor would need to specify the option based on whether the command was sent as a part of `compileOnSave` or similar.

File: src/server/protocol.ts
Comment (JoshuaKGoldberg) [reply src/server/protocol.ts:682]: Oh, sorry, I was a little confused on these - I'll change this one to be non-breaking.

File: src/harness/harnessLanguageService.ts
Comment (JoshuaKGoldberg) [inline src/harness/harnessLanguageService.ts:588]:
```suggestion

```

File: src/server/protocol.ts
Comment (JoshuaKGoldberg) [reply src/server/protocol.ts]: dacfd7e renames to `skipDestructiveCodeActions` and defaults it to falsy.

File: src/server/protocol.ts
Comment (amcasey) [inline src/server/protocol.ts:684]: I don't think we have an actual convention, but it looks like we have a pattern of using positive names and, if necessary, treating undefined as true.  e.g. `UserPreferences.includeAutomaticOptionalChainCompletions` or `CompilerOptions.AllowUnreachableCode`.  I think @andrewbranch has also found this frustrating and might have suggestions?

**Review by amcasey — APPROVED** [review]
I'd like to get a second opinion on the flag name, but LGTM otherwise.

File: src/server/protocol.ts
Comment (andrewbranch) [reply src/server/protocol.ts:684]: No, I don’t really have a preference here. This looks fine to me.

**Pull request merged**
